logstash-core/spec/logstash/java_pipeline_spec.rb

```diff
@@ -377,6 +377,110 @@ def flush(options)
+          sample_one( [{ "path" => {"to" => "Rome"}}] ) do
+            expect(subject).to be nil
+            expect(pipeline.last_error_evaluation_received).to match(/<no-class>:<null-value>/)
+          end
+        end
+
+        context "comparing incompatible types" do
+          sample_one( [{ "path" => {"to" => {"value" => [101, 102]}}}] ) do
+            expect(subject).to be nil
+            expect(pipeline.last_error_evaluation_received).to match(/Unexpected conditional input combination of.*List.*RubyFixnum/)
+          end
+        end
+      end
+
+      context "when the offending logic expression is used in a nested conditional structure" do
+        config <<-CONFIG
+                  filter {
+                    if "a" == "a" {
+                      if "b" == "b" {
+                        if [path][to][value] > 100 {
+                          mutate { add_tag => "hit" }
+                        } else {
+                          mutate { add_tag => "miss" }
+                        }
+                      }
+                    }
+                  }
+                CONFIG
+
+        sample_one( [{ "path" => {"to" => {"value" => "101"}}}] ) do
+          expect(subject).to be nil
+          expect(pipeline.last_error_evaluation_received).to match(/no implicit conversion of nil into Integer/)
+        end
+      end
+
+      context "when the offending condition is in the output section" do
+        before do
+          LogStash::PLUGIN_REGISTRY.add(:input, "spec_sampler_input", PipelineHelpers::SpecSamplerInput)
+          LogStash::PLUGIN_REGISTRY.add(:output, "spec_sampler_output", PipelineHelpers::SpecSamplerOutput)
+        end
+
+        describe "given a pipeline executing an event that would trigger an evaluation error" do
+          let(:pipeline) do
+            settings.set_value("queue.drain", true)
+            LogStash::JavaPipeline.new(
+              org.logstash.config.ir.PipelineConfig.new(
+                LogStash::Config::Source::Local, :main,
+                SourceWithMetadata.new(
+                  "config_string", "config_string",
+                  "input { spec_sampler_input {} }\n output { if [path][to][value] > 100 { spec_sampler_output {} } }"
+                ), settings
+              )
+            )
+          end
+          let(:event) do
+            [LogStash::Event.new({ "path" => {"to" => {"value" => "101"}}})]
+          end
+
+          let(:results) do
+            PipelineHelpers::SpecSamplerInput.set_event event
+            pipeline.run
+            PipelineHelpers::SpecSamplerOutput.seen
+          end
+
+          after do
+            pipeline.close
+          end
+
+          subject {results.length > 1 ? results : results.first}
+
+          it "should raise an error without killing the pipeline" do
+            expect(subject).to be nil
+            expect(pipeline.last_error_evaluation_received).to match(/no implicit conversion of nil into Integer/)
```

Comment:
This is a great work @andsel, thank you. 
I do believe this change also cover non-unicode payload case when using condition within the filter ([public issue](https://github.com/elastic/logstash/issues/15091)). Can we please add at least unit test for this case?

Another my concern, as my understanding with this change, we are _dropping_ the event if we have a conditional error. For the data sensitive ingestion systems this might not be desirable behavior, crash would actually indicate that the issue must be addressed to prevent data loss. So, I am thinking if we need to think about adding (or align on existing) option on event dropping decision.

Reply:
>  crash would actually indicate that the issue must be addressed to prevent data loss.

The goal of this PR is effectively to avoid pipeline crash. The event is logged, however there is another companion PR that will be created after this that DLQ the event in case DLQ is enable for this pipeline, so we don't loose the event.

Reply:
The test was added with f60892c5a029753dd61300c74dabe91157d14e6b